### PR TITLE
remove origin region range after applying a new version snapshot.

### PR DESCRIPTION
**Pull request by siddontang**
1 [a, c) -> 1 [a, b) + 2  [b, c)
If the follower applies the 1 [a, b) snapshot, we must remove origin `c -> 1` in region_ranges, otherwise, we may have two different keys mapping the same 1 `c->1` and `b -> 1` in region_ranges.

@ngaut @BusyJay @disksing @tiancaiamao

### CODE REVIEW

File: src/raftstore/store/peer_storage.rs [hunk 1]
```diff
@@ -306,15 +308,20 @@ impl PeerStorage {
...
-        try!(self.scan_region(self.engine.as_ref(),
-                              &mut |key, _| {
-                                  try!(w.delete(key));
-                                  Ok(true)
-                              }));
-        info!("clean old data takes {:?}", timer.elapsed());
-        timer = Instant::now();
+
+        if self.is_initialized() {
+            // we can only delete the old data when the peer is initialized.
```
Comment (ngaut) [inline src/raftstore/store/peer_storage.rs]: Why not just call destory_region ?

Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: we have no this function now.

File: tests/raftstore/test_split_region.rs [hunk 2]
```diff
@@ -302,4 +302,79 @@ fn test_node_split_overlap_snapshot() {
...
+
+    for _ in 0..100 {
+        // write many logs to force log GC for region 1 and region 2.
+        cluster.get(b"k1").unwrap();
+        cluster.get(b"k2").unwrap();
+    }
+
+    let region2 = pd_client.get_region(b"k2").unwrap();
+    // split [k2, +inf) -> [k2, k3), [k3, +inf)
+    cluster.must_split(&region2, b"k3");
```
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: Why split again?

Comment (siddontang) [inline tests/raftstore/test_split_region.rs]: For more test, and this covers the bug for applying snapshot with uninitialized peer.

File: src/raftstore/store/store.rs [hunk 3]
```diff
@@ -483,9 +483,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
 
+    fn on_ready_apply_snapshot(&mut self, apply_result: ApplySnapResult) {
+        let prev_region = apply_result.prev_region;
+        let region = apply_result.region;
+
+        info!("snapshot for region {:?} is applied ready", region);
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: s/ready//

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -483,9 +483,28 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+
+        if !prev_region.get_peers().is_empty() {
+            info!("region changed from {:?} -> {:?} after applying snapshot",
+                  prev_region,
+                  region);
+            // we have already initialized the peer, so must in region_ranges.
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: s/so/so it/

File: tests/raftstore/test_split_region.rs (hunk 2 above)
Comment (BusyJay) [inline tests/raftstore/test_split_region.rs]: Why can't region 2 cover this?

Comment (siddontang) [inline tests/raftstore/test_split_region.rs]: yep, it can cover too.

File: tests/raftstore/test_split_region.rs [hunk 5]
```diff
@@ -302,4 +302,65 @@ fn test_node_split_overlap_snapshot() {
...
+}
+
+fn test_apply_new_version_snapshot<T: Simulator>(cluster: &mut Cluster<T>) {
+    // truncate the log quickly so that we can force sending snapshot.
+    cluster.cfg.store_cfg.raft_log_gc_tick_interval = 20;
+    cluster.cfg.store_cfg.raft_log_gc_limit = 2;
+
+    // We use three nodes 1, 2, 3 for this test.
+    cluster.bootstrap_region().expect("");
+    cluster.start();
```
Comment (ngaut) [inline tests/raftstore/test_split_region.rs]: Can we merge those two function into a single function ? I saw the pattern many times.

Comment (siddontang) [inline tests/raftstore/test_split_region.rs]: yes, but should be in next PR.

File: tests/raftstore/test_split_region.rs [hunk 6]
```diff
@@ -302,4 +302,65 @@ fn test_node_split_overlap_snapshot() {
...
+    cluster.must_split(&region, b"k2");
+    cluster.must_put(b"k2", b"v2");
+
+    // node 1 and node 2 must have k2, but node 3 must not.
+    for i in 1..3 {
+        let engine = cluster.get_engine(i);
+        util::must_get_equal(&engine, b"k2", b"v2");
+    }
+
+    let engine = cluster.get_engine(3);
```
Comment (ngaut) [inline tests/raftstore/test_split_region.rs]: s/engine/engine3/

File: tests/raftstore/test_split_region.rs [hunk 7]
```diff
@@ -302,4 +302,65 @@ fn test_node_split_overlap_snapshot() {
...
-}
\ No newline at end of file
+}
+
+fn test_apply_new_version_snapshot<T: Simulator>(cluster: &mut Cluster<T>) {
+    // truncate the log quickly so that we can force sending snapshot.
+    cluster.cfg.store_cfg.raft_log_gc_tick_interval = 20;
+    cluster.cfg.store_cfg.raft_log_gc_limit = 2;
+
+    // We use three nodes 1, 2, 3 for this test.
```
Comment (ngaut) [inline tests/raftstore/test_split_region.rs]: s/nodes 1, 2, 3/nodes([1, 2, 3])/

File: src/raftstore/store/peer_storage.rs [hunk 8]
```diff
@@ -306,15 +308,20 @@ impl PeerStorage {
...
+            let timer = Instant::now();
+            // Delete everything in the region for this peer.
+            try!(self.scan_region(self.engine.as_ref(),
+                                  &mut |key, _| {
+                                      try!(w.delete(key));
+                                      Ok(true)
+                                  }));
+            info!("clean old data takes {:?}", timer.elapsed());
```
Comment (ngaut) [inline src/raftstore/store/peer_storage.rs]: s/data/region/

**Comment by ngaut** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**